components/raftstore-v2/src/operation/query/mod.rs

```diff
@@ -8,28 +8,75 @@
 //! first. Lease maintainance is implemented in lease module.
 //!
 //! Status query is implemented in the root module directly.
+//! Follower's read index is implemenented follower_read_index
+//! Leader's read index is implemented in read_index
+//! stale read check is implemented replica_read.rs
 
 use engine_traits::{KvEngine, RaftEngine};
 use kvproto::raft_cmdpb::{RaftCmdRequest, RaftCmdResponse, StatusCmdType};
 use raftstore::{
-    store::{cmd_resp, region_meta::RegionMeta, util, GroupState, ReadCallback},
+    store::{
+        cmd_resp, region_meta::RegionMeta, util, util::LeaseState, GroupState, ReadCallback,
+        RequestInspector, RequestPolicy,
+    },
     Error, Result,
 };
 use tikv_util::box_err;
+use txn_types::WriteBatchFlags;
 
 use crate::{
     fsm::PeerFsmDelegate,
     raft::Peer,
-    router::{DebugInfoChannel, QueryResChannel, QueryResult},
+    router::{DebugInfoChannel, QueryResChannel, QueryResult, ReadResponse},
 };
 
+mod follower_read_index;
 mod local;
+mod read_index;
+mod replica_read;
+
+impl<'a, EK: KvEngine, ER: RaftEngine, T: raftstore::store::Transport>
+    PeerFsmDelegate<'a, EK, ER, T>
+{
+    fn inspect_read(&mut self, req: &RaftCmdRequest) -> Result<RequestPolicy> {
+        let flags = WriteBatchFlags::from_bits_check(req.get_header().get_flags());
+        if flags.contains(WriteBatchFlags::STALE_READ) {
+            return Ok(RequestPolicy::StaleRead);
+        }
+
+        match self.fsm.peer_mut().inspect_lease() {
+            LeaseState::Valid => Ok(RequestPolicy::ReadLocal),
+            LeaseState::Expired | LeaseState::Suspect => {
+                // Perform a consistent read to Raft quorum and try to renew the leader lease.
+                Ok(RequestPolicy::ReadIndex)
+            }
+        }
+    }
 
-impl<'a, EK: KvEngine, ER: RaftEngine, T> PeerFsmDelegate<'a, EK, ER, T> {
     #[inline]
     pub fn on_query(&mut self, req: RaftCmdRequest, ch: QueryResChannel) {
         if !req.has_status_request() {
-            unimplemented!();
+            let mut resp = RaftCmdResponse::default();
+            let term = self.fsm.peer().term();
+            cmd_resp::bind_term(&mut resp, term);
+            let policy = self.inspect_read(&req);
+            match policy {
+                Ok(RequestPolicy::ReadIndex) => {
+                    self.fsm
+                        .peer_mut()
+                        .read_index(self.store_ctx, req, resp, ch);
+                }
+                Ok(RequestPolicy::ReadLocal) => {
+                    let read_resp = ReadResponse::new(0);
```

Comment:
Why returns 0 here? I guess it just notify the sender that the leader is in lease now, so the sender can retry the read using LocalReader again?

Reply:
> Why returns 0 here? I guess it just notifies the sender that the leader is in lease now, so the sender can retry the read using LocalReader again?

Yes. Generally speaking I think as long as it returns QueryResult::Read() regardless the read index value, then it means the local read can be done.